Lib/functools.py

```diff
@@ -826,3 +826,32 @@ def wrapper(*args, **kw):
     wrapper._clear_cache = dispatch_cache.clear
     update_wrapper(wrapper, func)
     return wrapper
+
+
+################################################################################
+### cached_property() - computed once per instance, cached as attribute
+################################################################################
+
+class cached_property:
+    def __init__(self, func):
+        self.func = func
+        self.__doc__ = func.__doc__
+        self.lock = RLock()
+
+    def __get__(self, instance, cls=None):
```

Comment:
When `cls` is `None`? When `__get__` is called without `cls` parameter?
Btw in docs for descriptors, the third argument is called `owner` IIRC.

Reply:
AFAIK every call to `__get__` generated by attribute lookup machinery will include the second (not including `self`) argument, but both the descriptor documentation (https://docs.python.org/3/reference/datamodel.html#descriptors) and the descriptor howto guide (https://docs.python.org/3/howto/descriptor.html) explicitly include an example of manually calling a descriptor's `__get__` method providing only one argument. Also, all of the code examples in the HOWTO show the second argument as optional. Given that the method _can_ be called directly and `cls` is unused in this implementation, it seems reasonable to follow that lead and make it optional.

Regarding the argument name, considering both the Data Model documentation and the HOWTO, we see examples in Python's own documentation of descriptors whose `__get__` second argument is named `owner`, `type`, or `objtype`. All three of these are found in the standard library, along with `cls`, `klass`, `ownerclass`, and `_type`. There doesn't seem to be a standard usage. Personally I find `cls` or `klass` or `objtype` clearer, as `owner` doesn't clarify the key difference from the other argument, which is that it's a type. Perhaps naming the two arguments `obj` and `objtype` does the best job of clearly indicating their relationship. That said, I don't feel strongly and will rename to `owner` if you prefer that.

Reply:
I doubt if howto is 100% correct.

*datamodel* doesn't allow `None` for `owner`, AFAIK Python aways pass a class type.

Using default for always specified argument confuses me.

*datamodel* uses `owner`, please do it as well.